src/libcollections/string.rs

```diff
@@ -29,7 +36,185 @@ use str::{self, FromStr, Utf8Error, Chars};
+//!
+//! ```
+//! // some bytes, in a vector
+//! let sparkle_heart = vec![240, 159, 146, 150];
+//!
+//! // We know these bytes are valid, so we'll use `unwrap()`.
+//! let sparkle_heart = String::from_utf8(sparkle_heart).unwrap();
+//!
+//! assert_eq!("💖", sparkle_heart);
+//! ```
+//!
+//! [`from_utf8()`]: #method.from_utf8
+//!
+//! # UTF-8
+//!
+//! `String`s are always valid UTF-8. This has a few implications, the first of
+//! which is that if you need a non-UTF-8 string, consider [`OsString`]. It is
+//! similar, but without the UTF-8 constraint. The second implication is that
+//! you cannot index into a `String`:
+//!
+//! ```ignore
+//! let s = "hello";
+//!
+//! println!("The first letter of s is {}", s[0]); // ERROR!!!
+//! ```
+//!
+//! [`OsString`]: ../ffi/struct.OsString.html
+//!
+//! Indexing is intended to be a constant-time operation, but UTF-8 encoding
+//! does not allow us to do this. Furtheremore, it's not clear what sort of
+//! thing the index should return: a byte, a codepoint, or a grapheme cluster.
+//! The [`as_bytes()`] and [`chars()`] methods return iterators over the first
+//! two, respectively.
+//!
+//! [`as_bytes()`]: #method.as_bytes
+//! [`chars()`]: #method.chars
+//!
+//! # Deref
+//!
+//! `String`s implement [`Deref`]`<Target=str>`, and so inherit all of [`str`]'s
+//! methods. In addition, this means that you can pass a `String` to any
+//! function which takes a [`&str`] by using an ampersand (`&`):
+//!
+//! ```
+//! fn takes_str(&str) { }
+//!
+//! let s = String::from("Hello");
+//!
+//! takes_str(&s);
+//! ```
+//!
+//! [`&str`]: ../primitive.str.html
+//! [`Deref`]: ../ops/trait.Deref.html
+//!
+//! This will create a [`&str`] from the `String` and pass it in. This
+//! conversion is very inexpensive, and so generally, functions will accept
+//! [`&str`]s as arguments unless they need a `String` for some specific reason.
+//!
+//!
+//! # Representation
+//!
+//! A `String` is made up of three components: a pointer to some bytes, a
+//! length, and a capacity. The pointer points to an internal buffer `String`
+//! uses to store its data. The length is the number of elements currently
+//! stored in the buffer, and the capacity is the size of the buffer. As such,
+//! the length will always be less than or equal to the capacity.
+//!
+//! This buffer is always stored on the heap.
+//!
+//! You can look at these with the [`as_ptr()`], [`len()`], and [`capacity()`]
+//! methods:
+//!
+//! ```
+//! let story = String::from("Once upon a time...");
+//!
+//! let ptr = story.as_ptr();
+//! let len = story.len();
+//! let capacity = story.capacity();
+//!
+//! // story has thirteen bytes
+//! assert_eq!(19, len);
+//!
+//! // We can re-build a String out of ptr, len, and capacity. This is all
+//! // unsafe becuase we are responsible for making sure the components are
+//! // valid:
+//! let s = unsafe { String::from_raw_parts(ptr, len, capacity) } ;
+//!
+//! assert_eq!(s, story);
+//! ```
```

Comment:
This example results in two owners of the buffer. It should crash with a double-free...